x-pack/legacy/plugins/apm/server/lib/service_map/setup_required_scripts.ts

```diff
@@ -0,0 +1,298 @@
+}
+
+async function putScriptReduceServiceConns(callCluster: CallCluster) {
+  return await callCluster('putScript', {
+    id: REDUCE_SERVICE_CONNS_SCRIPT_ID,
+    body: {
+      script: {
+        lang: 'painless',
+        source: `
+          void extractChildren(def caller, def spans, def upstream, def conns, def count) {
+            // TODO: simplify this
+            if (spans.containsKey(caller.id)) {
+              for(s in spans[caller.id]) {
+                  if (caller.span_type=='external') {
+                    upstream.add(caller.service_name+"/"+caller.environment);
+
+                    def conn = new HashMap();
+                    conn.caller = caller;
+                    conn.callee = s;
+                    conn.upstream = new ArrayList(upstream);
+                    conns.add(conn);
+
+                    extractChildren(s, spans, upstream, conns, count);
+                    upstream.remove(upstream.size()-1);
+                  } else {
+                    extractChildren(s, spans, upstream, conns, count);
+                  }
+              }
+            } else {
+              // no connection found
+              def conn = new HashMap();
+              conn.caller = caller;
+              conn.upstream = new ArrayList(upstream);
+              conn.upstream.add(caller.service_name+"/"+caller.environment);
+              conns.add(conn);
+            }
+          }
+          def conns = new HashSet();
+          def spans = new HashMap();
+
+          // merge results from shards
+          for(state in states) {
+            for(s in state.entrySet()) {
+              def v = s.getValue();
+              def k = s.getKey();
+              if(!spans.containsKey(k)) {
+                spans[k] = v;
+              } else {
+                for (p in v) {
+                  spans[k].add(p);
+                }
+              }
+            }
+          }
+
+          if (spans.containsKey(null) && spans[null].size() > 0) {
+            def node = spans[null][0];
+            def upstream = new ArrayList();
+
+            extractChildren(node, spans, upstream, conns, 0);
+
+            return new ArrayList(conns)
+          }
+          return [];
+        `
+      }
+    }
+  });
+}
+
+async function putScriptCombineServiceConns(callCluster: CallCluster) {
+  return await callCluster('putScript', {
+    id: COMBINE_SERVICE_CONNS_SCRIPT_ID,
+    body: {
+      script: {
+        lang: 'painless',
+        source: `return state.spans`
+      }
+    }
+  });
+}
+
+async function putIngestPipelineExtractDestination(callCluster: CallCluster) {
+  return await callCluster('ingest.putPipeline', {
+    id: EXTRACT_DESTINATION_INGEST_PIPELINE_ID,
+    body: {
+      description: 'sets destination on ext spans based on their name',
+      processors: [
+        {
+          set: {
+            if: "ctx.span != null && ctx.span.type == 'ext'",
+            field: 'span.type',
+            value: 'external'
+          }
```

Comment:
Looks like we rewrite `ext` to `external`. Why is that?

Reply:
At the time, the value was not consistent, so this normalized everything to `external`.

see https://github.com/elastic/apm-agent-nodejs/issues/1225#issue-469295459

Reply:
should also note that this ingest pipeline will not be merged.